dialog_widget: Add optional parameter `update_submit_on_state_change`.

Ddharmani3 opened the pull request:
<!-- Describe your pull request here.-->
Add an optional dialog_widget parameter that takes a boolean value. If set to true, it activates the "Confirm" button on a state change of any input field in the modal and also disables it again if the input field value is changed back to the original value.

Implementation: Store all the originalValues of input fields in the modal once it is rendered and add an event listener to all the input fields that listen for input and check if the new value is different from the original value. It activates the "Confirm" button if the new value is different, and makes it disabled if it is the same.

In this PR, I have only used the new option for the email change modal. I will soon create an Issue as a follow-up to this PR, that will use it everywhere else, such as the bot edit modal and move topic modal, as mentioned [here](https://github.com/zulip/zulip/issues/22683#issuecomment-1230805433).

Also, I will fix the cursor to be at the end of the input on focus for the email input field and other input fields where it is needed once I create a focus_on_text_input function, as described [here](https://github.com/zulip/zulip/pull/24823#discussion_r1147433110).

Fixes: <!-- Issue link, or clear description.-->
#22683
<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

https://user-images.githubusercontent.com/64723994/228642612-53db87d4-4930-4e85-b1c0-0203d965bc7f.mp4

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
How will this work if a dialog widget has multiple inputs?

ganpa3 commented:
@Ddharmani3 Will this work for other inputs such as `select`, `textarea`, etc.?

Ddharmani3 commented:
> @Ddharmani3 Will this work for other input elements such as `select`, `textarea`, etc.?

@ganpa3 yes it will work for input elements such as `select`, `textarea`, `checkboxes`, etc. after making some changes as mentioned [here](https://github.com/zulip/zulip/pull/24911#issuecomment-1489673969).In `$(".modal__content :input")` using :input will make it work for inputs, textarea, select, etc except checkboxes as documented [here](https://api.jquery.com/input-selector/). We can also make it work for checkboxes if needed.

ganpa3 commented on web/src/dialog_widget.js:
```diff
@@ -118,6 +119,24 @@ export function launch(conf) {
+        const originalValues = {};
+        $(() => {
+            $submit_button.prop("disabled", true);
+            $(".modal__content input").each(function () {
```
So, shouldn't this be `:input`?

Ddharmani3 replied:
Yes, but in this PR, I have only applied this for the Email change modal, which has only one input field, and its input type is covered by the current `$(".modal__content input")` selector. In the follow-up PR, where I will apply this option to more modals, I will update `$(".modal__content input")` to cover the input types present in those modals as per requirements.

or Should I update $(".modal__content input") to accept all input field types in this PR only?"

Ddharmani3 commented:
@ganpa3 @sahil839, I have updated the PR to include all input field types by using $(".modal__content :input"). Additionally, I have addressed a bug discovered by @sahil839 [here](https://github.com/zulip/zulip/pull/24910#issuecomment-1490067588): In a modal with multiple input fields, if we change one input, change the second input, and then change any of the two fields back to the original value, the submit button is disabled even though one value is still changed from the original value. I have fixed this bug in the updated PR. Now, the submit button will stay disabled if all the values are the same as the original values, and it will be activated if one or more values are different than the original.

sahil839 commented on web/src/dialog_widget.js:
```diff
@@ -118,6 +119,29 @@ export function launch(conf) {
     const $submit_button = $dialog.find(".dialog_submit_button");
 
+    if (conf.update_submit_on_state_change) {
+        const originalValues = {};
```
We generally don't use camelCase variable names and instead use names separated by underscores.

sahil839 commented on web/src/dialog_widget.js:
```diff
@@ -118,6 +119,29 @@ export function launch(conf) {
+        $(() => {
+            $submit_button.prop("disabled", true);
+            $(".modal__content :input").each(function () {
+                originalValues[$(this).attr("name")] = currentValues[$(this).attr("name")] =
```
I think it would be much more readable if we use a variable like `property_name` for `$(this).attr("name")` and also set the values for the two objects in separate lines.

Ddharmani3 replied:
> I think it would be much more readable if we use a variable like `property_name` for `$(this).attr("name")` and also set the values for the two objects in separate lines.

can we use `field_name` and `field_value`  or `input_name` and `input_value` for `$(this).attr("name")` and `$(this).val()`?

sahil839 reviewed (commented):
Posted a couple of comments.

This works for the modal that uses the added parameter in this PR, but may require some changes for a couple of inputs like the `dropdown_list_widget`.

sahil839 commented:
@Ddharmani3 can you rebase the PR?

Ddharmani3 commented:
> @Ddharmani3 can you rebase the PR?

@sahil839 resolved conflicts and rebased. PR is ready for review now.

sahil839 commented on web/src/dialog_widget.ts:
```diff
@@ -152,6 +155,34 @@ export function launch(conf: WidgetConfig): void {
+        const original_values: {[key: string]: unknown} = {};
+        const current_values: {[key: string]: unknown} = {};
+
+        $(() => {
```
Is it necessary to include the code below inside `$()`? The click handler is defined normally below though.

Ddharmani3 replied:
it checks that if the document is ready then it attaches click handlers to rendered input fields. not including this was causing errors sometimes as the code was trying to attach the click handlers to elements that were not present in DOM.

I have checked the updated code just now by removing this it is not giving an error but should we keep this to ensure that DOM is ready?

Ddharmani3 replied:
@sahil839 I will also update the commit message and push the code as soon as you confirm this.

sahil839 replied:
Ok, let's keep it as it is for now and can update if required after feedback from others.

Ddharmani3 replied:
As I mentioned above:
> I have checked the updated code just now by removing this it is not giving an error anymore

Hence, I removed `$()` in the updated PR after testing it several times. It is working properly every time. Do share your feedback in case you find the issue in your manual testing.

sahil839 commented:
The commit message for first commit uses the old variable names which are now changed.

Ddharmani3 commented:
> The commit message for first commit uses the old variable names which are now changed.

Update the commit message. PR is ready for review now.

sahil839 commented:
Looks good to me.

Ddharmani3 commented:
@sahil839 updated the commit message. PR is ready for review now.

sahil839 commented:
Cool. I do not have any more feedback. @timabbott this is ready for your review.

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -133,6 +134,8 @@ export function launch(conf: WidgetConfig): void {
     // * loading_spinner: Whether to show a loading spinner inside the
     //   submit button when clicked.
+    // * update_submit_on_state_change: If true, updates state of submit button
+    // on valid input change in modal
```
The `on` second line should be intended two characters, like in the previous option. Please pay attention to details like this when making changes -- it's your responsibility to write code consistent with the existing style, and it's time-consuming and slows review of your work if one can't count on you to do so.

timabbott commented:
This lgtm aside from the two comments posted above, thanks for your work on this @Ddharmani3.

Ddharmani3 commented:
@timabbott, addressed the [feedback](https://github.com/zulip/zulip/pull/24911#discussion_r1169374262). This PR is ready for review now.

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -152,6 +155,32 @@ export function launch(conf: WidgetConfig): void {
+        const current_values: {[key: string]: unknown} = {};
+
+        $submit_button.prop("disabled", true);
+        $(".modal__content :input").each(function () {
```
Can we fix this selector to apply just to modal__content inside this dialog widget? In theory we can't have other modals open, but I don't like hardcoding that in here -- it's the kind of thing that breaks if we change our assumptions in the future.

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -133,6 +134,8 @@ export function launch(conf: WidgetConfig): void {
     // * loading_spinner: Whether to show a loading spinner inside the
     //   submit button when clicked.
+    // * update_submit_on_state_change: If true, updates state of submit button
+    //   on valid input change in modal
```
This should end with a `.`, like all the others.

timabbott replied:
Let's also call this `update_submit_disabled_state_on_change` to be super explicit what exactly is being updated.

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -152,6 +155,32 @@ export function launch(conf: WidgetConfig): void {
+                // Check if the new value is different from the current value
+                if ($(this).val() !== current_values[property_name]) {
+                    current_values[property_name] = $(this).val();
+                }
```
Could we make this more robust by having `current_values` be read from the DOM each time? 

I worry about missing an event that changes the value but isn't an `input` event (e.g. some live-update) resulting in the logic being wrong. I think the best way to do this would be to just write a function that loops through the inputs and collects their values, and call that function in two places -- when the input is first rendered and again when there's a change to check if it's different.

Ddharmani3 commented:
@timabbott updated the PR to address feedback. This PR is ready for review now.

timabbott commented:
So the code here is now great, but it's time for you to start working on your commit messages.  https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html covers this in detail, but just looking at your second commit, it has the rough formatting of a good commit message, but the content is really not what we're looking for. 

First, the title doesn't explain very clearly at all the product change. And then the body spends all its time repeating what's obvious from glancing at the code -- which function call was changed and where -- which is essentially useless.

```
    settings: Update state of "Change" button on email change.
    
    Pass `update_submit_disabled_state_on_change` parameter as true to
    dialog_widget.launch() for change_email_modal to handle "Change"
    button's disable state on input field state change.
```

timabbott commented on web/src/dialog_widget.ts:
```diff
@@ -139,6 +142,33 @@ export function launch(conf: WidgetConfig): void {
+        $inputs.on("input", () => {
+            const current_values = get_current_values($inputs);
+
+            if (JSON.stringify(original_values) !== JSON.stringify(current_values)) {
```
Hmm, actually, this comparison doesn't make any sense -- JSON stringification is not necessarily stable if there's multiple items in the objects. Changed to `_.isEqual`, which is an actual function for deep equality comparisons of objects.

timabbott commented:
Merged after rewriting the commit messages. @Ddharmani3 please learn from the changes I made, and also take some time to read commit messages from other core team members -- this is one of the most important parts of contributing to larger projects and I think you're ready to focus on learning how to do it well.
